pkg/glusterfs/controllerserver.go

```diff
@@ -0,0 +1,377 @@
+	*GfDriver
+}
+
+func (cs *ControllerServer) CreateVolume(ctx context.Context, req *csi.CreateVolumeRequest) (*csi.CreateVolumeResponse, error) {
+
+	var glusterServer string
+	var bkpServers []string
+
+	if req == nil {
+		glog.Errorf("volume create request is nil")
+		return nil, status.Errorf(codes.InvalidArgument, "request cannot be empty")
+	}
+
+	if req.Name == "" {
+		return nil, status.Error(codes.InvalidArgument, "Name is a required field")
+	}
+	glog.V(1).Infof("creating volume with name ", req.Name)
+
+	if req.VolumeCapabilities == nil || len(req.VolumeCapabilities) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "Volume capabilities is a required field")
+	}
+
+	// If capacity mentioned, pick that or use default size 20 MB
+	volSizeBytes := defaultVolumeSizeInMB
+	if req.GetCapacityRange() != nil {
+		volSizeBytes = int(req.GetCapacityRange().GetRequiredBytes())
+	}
+	volSizeMB := int((volSizeBytes + MB - 1) / MB)
+
+	// Get Volume name : TODO use the values from request
+	volumeName := req.Name
+	glusterVol := req.GetParameters()["glustervol"]
+	glusterServer = req.GetParameters()["glusterserver"]
+	glusterURL := req.GetParameters()["glusterurl"]
+	glusterURLPort := req.GetParameters()["glusterurlport"]
+	glusterUser := req.GetParameters()["glusteruser"]
+	glusterUserSecret := req.GetParameters()["glusterusersecret"]
+
+	glog.V(3).Infof("Request fields:[ %v %v %v %v %v %v]", glusterVol, glusterServer, glusterURL, glusterURLPort, glusterUser, glusterUserSecret)
+
+	// Get list of volumes in the TSP
+	volumeListResp, err := cs.client.Volumes("")
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	glog.V(2).Infof("Volume list response: %+v", volumeListResp)
+
+	/* TODO: Check the requested volume exist in the cluster
+	glusterServer, bkpServers, err = cs.checkExistingVolumes(&volumeListResp, volumeName, volSizeBytes)
+	if err == nil {
+		return &csi.CreateVolumeResponse{
+			Volume: &csi.Volume{
+				Id:            volumeName,
+				CapacityBytes: int64(volSizeBytes),
+				Attributes: map[string]string{
+					"glustervol":        volumeName,
+					"glusterserver":     glusterServer,
+					"glusterbkpservers": strings.Join(bkpServers, ":"),
+				},
+			},
+		}, nil
+	}
+	*/
+
+	// If volume does not exist, provision volume
+	glog.V(4).Infof("Received request to create/provision volume name:%s with size:%d", volumeName, volSizeMB)
+	volMetaMap := make(map[string]string)
+	volMetaMap[glusterDescAnn] = glusterDescAnnValue
+	volumeReq := api.VolCreateReq{
+		Name:     volumeName,
+		Metadata: volMetaMap,
+		Size:     uint64(volSizeMB),
+	}
+
+	glog.V(2).Infof("volume request: %+v", volumeReq)
+	volumeCreateResp, err := cs.client.VolumeCreate(volumeReq)
+	if err != nil {
+		glog.Errorf("failed to create volume : %v", err)
+		return nil, status.Error(codes.Internal, fmt.Sprintf("failed to create volume: %s", err.Error()))
+	}
+
+	glog.V(3).Infof("volume create response : %+v", volumeCreateResp)
```

Comment:
There should be distinction between info and debug levels. This entry dumps the whole response